contracts/MuonNodeManager.sol

```diff
@@ -0,0 +1,354 @@
+
+    // commit_id => git commit id
+    mapping(string => string) public configs;
+
+    uint64 public lastRoleId;
+
+    // hash(role) => role id
+    mapping(bytes32 => uint64) public roleIds;
+
+    // role id => node id => index + 1
+    mapping(uint64 => mapping(uint64 => uint16)) public nodesRoles;
+
+    /**
+     * @dev Modifier to update the lastUpdateTime state variable.
+     */
+    modifier updateState() {
+        lastUpdateTime = block.timestamp;
+        _;
+    }
+
+    /**
+     * @dev Modifier to update the lastEditTime of a specific node.
+     * @param nodeId The id of the node.
+     */
+    modifier updateNodeState(uint64 nodeId) {
+        nodes[nodeId].lastEditTime = block.timestamp;
+        _;
+    }
+
+    function __MuonNodeManagerUpgradeable_init() internal initializer {
+        __AccessControl_init();
+
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _setupRole(ADMIN_ROLE, msg.sender);
+        _setupRole(DAO_ROLE, msg.sender);
+
+        lastNodeId = 0;
+        lastUpdateTime = block.timestamp;
+        lastRoleId = 0;
+    }
+
+    /**
+     * @dev Initializes the contract.
+     */
+    function initialize() external initializer {
+        __MuonNodeManagerUpgradeable_init();
+    }
+
+    function __MuonNodeManagerUpgradeable_init_unchained()
+        internal
+        initializer
+    {}
+
+    /**
+     * @dev Adds a new node.
+     * Only callable by the ADMIN_ROLE.
+     * @param _nodeAddress The address of the node.
+     * @param _stakerAddress The address of the staker associated with the node.
+     * @param _peerId The peer ID of the node.
+     * @param _active Indicates whether the node is active or not.
+     */
+    function addNode(
+        address _nodeAddress,
+        address _stakerAddress,
+        string calldata _peerId,
+        bool _active
+    )
+        public
+        override
+        onlyRole(ADMIN_ROLE)
+        updateState
+    {
+        require(nodeAddressIds[_nodeAddress] == 0, "Node address is already registered.");
+
+        require(
+            stakerAddressIds[_stakerAddress] == 0,
+            "Staker address is already registered."
+        );
+
+        lastNodeId++;
+        nodes[lastNodeId] = Node({
+            id: lastNodeId,
+            nodeAddress: _nodeAddress,
+            stakerAddress: _stakerAddress,
+            peerId: _peerId,
+            tier: 0,
+            active: _active,
+            roles: new uint64[](0),
+            startTime: block.timestamp,
+            lastEditTime: block.timestamp,
+            endTime: 0
+        });
+
+        nodeAddressIds[_nodeAddress] = lastNodeId;
+        stakerAddressIds[_stakerAddress] = lastNodeId;
+
+        emit NodeAdded(lastNodeId, nodes[lastNodeId]);
+    }
+
+    /**
+     * @dev Allows the admins to deactivate the nodes.
+     * Only callable by the ADMIN_ROLE.
+     * @param nodeId The ID of the node to be deactivated.
+     */
+    function deactiveNode(uint64 nodeId)
```

Comment:
We need a function to let the DAMIN activate a node. Otherwise, a node could be deactivated and its staked amount will be locked

Reply:
Let's allow the admin to decative a node via staking contract